File: Python/DoublyLinkedList/DoublyLinkedList.py
class Node:
  def __init__(self,data):
    self.data = data
    self.prev = None
    self.next = None

class DoublyLinkedList:
  def __init__(self, data = None)-> None:
    if data is None:
      self.head = None
      self.tail = None
      self.length = 0
    else:
      newNode = Node(data)
      self.head = newNode
      self.tail = newNode
      self.length = 1

  def printList(self)->None:
    if self.length == 0:
      print("Empty List")
      return
    temp = self.head
    for i in range(self.length):
      print(temp.data)
      temp = temp.next
  def append(self,data)->None:
    newNode = Node(data)
    if self.length == 0:
      self.head = newNode
      self.tail = newNode
    else:
      self.tail.next = newNode
      newNode.prev = self.tail
      self.tail = newNode
    self.length += 1
    
  def deleteLast(self)->None:
    if self.length == 0:
      return
    temp = self.tail
    if self.length == 1:
      self.head = None
      self.tail = None
    else:
      self.tail = self.tail.prev
      self.tail.next = None
    del(temp)
    self.length -= 1
    
  def prepend(self,data)->None:
    newNode = Node(data)
    if self.length == 0:
      self.head = newNode
      self.tail = newNode
    else:
      newNode.next = self.head
      self.head.prev = newNode
      self.head = newNode
    self.length += 1
    
  def deleteFirst(self)->None:
    if self.length == 0:
      return
    temp = self.head
    if self.length == 1:
      self.head = None
      self.tail = None
    else:
      self.head = self.head.next
      self.head.prev = None
    del(temp)
    self.length -= 1
    
  def get(self, index)->Node:
    if index < 0 or index >= self.length:
      # raise ValueError("Index out of bound for function \"get\"")
      return
    temp =  self.head
    for i in range(index):
      temp = temp.next
    return temp
  
  def set(self,index,data)->bool:
    if index < 0 or index >= self.length:
      # raise ValueError("Index out of bound for function \"set\"")
      return False
    temp = self.get(index)
    temp.data = data
    return True
  
  def insert(self, index, data)->bool:
    if index < 0 or index > self.length:
      # raise ValueError("Index out of bound for function \"set\"")
      return False
    if index == 0:
      self.prepend(data)
      return True
    if index == self.length:
      self.append(data)
      return True
    newNode = Node(data)
    temp = self.get(index)
    newNode.next = temp
    newNode.prev = temp.prev
    temp.prev.next = newNode
    temp.prev = newNode
    self.length += 1
    return True
    
  def deleteNode(self, index):
    if index < 0 or index >= self.length:
      # raise ValueError("Index out of bound for function \"set\"")
      return
    if index == 0:
      self.deleteFirst()
      return
    if index == self.length - 1:
      self.deleteLast()
      return
  
    temp = self.get(index)
    temp.prev.next = temp.next
    temp.next.prev = temp.prev
    del(temp)
    self.length -= 1 

File: C++/DoublyLinkedList/DoublyLinkedList.cpp
#include "DoublyLinkedList.h"
#include <iostream>
using namespace std;

Node::Node(int data)
{
  this->data = data;
  prev = nullptr;
  next = nullptr;
}

DoublyLinkedList::DoublyLinkedList(int data)
{
  Node *newNode = new Node(data);
  head = newNode;
  tail = newNode;
  length = 1;
}

void DoublyLinkedList::append(int data)
{
  Node *newNode = new Node(data);
  if (length == 0)
  {
    head = newNode;
    tail = newNode;
  }
  else
  {
    newNode->prev = tail;
    tail->next = newNode;
    tail = newNode;
  }
  length++;
}

void DoublyLinkedList::deleteLast()
{
  if (length == 0)
    return;
  Node *temp = tail;
  if (length == 1)
  {
    head = nullptr;
    tail = nullptr;
  }
  else
  {
    tail = tail->prev;
    tail->next = nullptr;
  }
  delete (temp);
  length--;
}

void DoublyLinkedList::prepend(int data)
{
  Node *newNode = new Node(data);
  if (length == 0)
  {
    head = newNode;
    tail = newNode;
  }
  else
  {
    newNode->next = head;
    head->prev = newNode;
    head = newNode;
  }
  length++;
}

void DoublyLinkedList::deleteFirst()
{
  if (length == 0)
    return;
  Node *temp = head;
  if (length == 1)
  {
    head = nullptr;
    tail = nullptr;
  }
  else
  {
    head = head->next;
    head->prev = nullptr;
  }
  delete (temp);
  length--;
}

Node *DoublyLinkedList::get(int index)
{
  if (index < 0 || index >= length)
  {
    cout << "Invalid index for method \"get\"" << endl;
    return nullptr;
  }

  Node *temp = head;
  for (int i = 0; i < index; i++)
    temp = temp->next;

  return temp;
}

bool DoublyLinkedList::set(int index, int data)
{
  if (index < 0 || index >= length)
  {
    cout << "Invalid index for method \"set\"" << endl;
    return false;
  }

  Node *temp = head;
  for (int i = 0; i < index; i++)
    temp = temp->next;
  temp->data = data;
  return true;
}

bool DoublyLinkedList::insert(int index, int value)
{
  if (index < 0 || index > length)
  {
    cout << "Invalid index for method \"set\"" << endl;
    return false;
  }
  if (index == 0)
  {
    prepend(value);
    return true;
  }
  if (index == length)
  {
    append(value);
    return true;
  }
  Node *newNode = new Node(value);
  Node *temp = get(index);
  newNode->prev = temp->prev;
  newNode->next = temp;
  temp->prev->next = newNode;
  temp->prev = newNode;
  length++;
  return true;
}

void DoublyLinkedList::deleteNode(int index)
{
  if (index < 0 || index >= length)
  {
    cout << "Invalid index for method \"deleteNode\"" << endl;
    return;
  }
  if (index == 0)
  {
    deleteFirst();
    return;
  }
  if (index == length - 1)
  {
    deleteLast();
    return;
  }
  Node *temp = get(index);
  temp->prev->next = temp->next;
  temp->next->prev = temp->prev;
  delete (temp);
  length--;
}